add support for running tests in parallel

boneskull opened the pull request:
> (this PR depends on most other PRs linked to #4198, so they should be merged first; documentation will be in another PR)

## Summary

This PR adds support for running test files in parallel via `--parallel`.  For many cases, this should "just work."

## How It Works

When the `--parallel` flag is supplied, Mocha will swap out the default `Runner` (`lib/runner.js`) for `BufferedRunner` (`lib/buffered-runner.js`).

`BufferedRunner` _extends_ `Runner`.  `BufferedRunner#run()` is the main point of extension.  Instead of executing the tests in serial, it will create a pool of worker processes (not worker _threads_) based on the maximum job count (`--jobs`; defaults to `<number of CPU cores> - 1`).  Both `BufferedRunner` and the `worker` module consume the abstraction layer, [workerpool](https://npm.im/workerpool).

`BufferedRunner#run()` does _not_ load the test files, unlike `Runner#run()`.  Instead, it has a list of test files, and puts these into an async queue.  The `EVENT_RUN_BEGIN` event is then emitted.  As files enter the queue, `BufferedRunner#run()` tells `workerpool` to execute the `run()` function of the pool.  `workerpool` then launches as many worker processes are needed--up to the maximum--and executes the `run()` function with a single filepath and any options for a `Mocha` instance.

The first time `lib/worker.js` is invoked, it will "bootstrap" itself, by handling `--require`'d modules and validating the UI.  Note that _reporter validation_ does not occur.  Once bootstrapped, it instantiate `Mocha`, add the single file, swap any reporter out for the `Buffered` reporter (`lib/reporters/buffered.js`) then execute `Mocha#run()`, which invokes `Runner#run()`.

The `Buffered` reporter listens for events emitting from the `Runner` instance, like a reporter usually does.  But instead of outputting to the console, it buffers the events in a queue.  Once the file has completed running, the queue is drained: the events collected are (trivially) serialized for transmission back to the main process.  `BufferedRunner#run()` receives the list of events, trivially _deserializes_ them, and re-emits the events to whatever the chosen reporter is (e.g., the `spec` reporter).  In this way, the reporters don't know that the tests were run in parallel.  Practically, the user will see reporter output in "chunks" instead of the "stream" of results they usually expect.  This method ensures that while the test files run in a nondeterministic order, the reporter output will be deterministic for any given test file.

Once the result (the queue of events) has been returned to the main process, the worker process stays open, but waits for further instruction.  If there are more files in `BufferedRunner#run()`'s queue, `workerpool` will instruct the worker to take the next file from the list, and so on, and so forth.  When all files have executed, the pool terminates, the `EVENT_RUN_END` event is emitted, and the reporter handles it.

## Limitations & Caveats

> (this section is pasted from the documentation with minimal edits)

### Reporter Limitations

Due to the nature of the following reporters, they cannot work when running tests in parallel:

- `markdown`
- `progress`
- `json-stream`

These reporters expect Mocha to know _how many tests it plans to run_ before execution. This information is unavailable in parallel mode, as test files are loaded only when they are about to be run.

In serial mode, tests results will "stream" as they occur. In parallel mode, reporter output is _buffered_; reporting will occur after each file is completed. In practice, the reporter output will appear in "chunks" (but will otherwise be identical).

### Order is Non-Deterministic

In parallel mode, we have no guarantees about the order in which test files will be run--or what process runs them--as it depends on the execution times of the test files.

Because of this, the following options _cannot be used_ in parallel mode:

- `--file`
- `--sort`
- `--delay`

### Test Duration Variability

Because running tests in parallel mode uses more system resources at once, the OS may take extra time to schedule and complete some operations. For this reason, test timeouts may need to be increased either globally or otherwise.

### "Bail" is "Best Effort"

When used with `--bail` (or `this.bail()`) to exit after the first failure, it's likely other tests will be running at the same time. Mocha must shut down its worker processes before exiting.

Likewise, subprocesses may throw uncaught exceptions. When used with `--allow-uncaught`, Mocha will "bubble" this exception to the main process, but still must shut down its processes.

### Root Hooks Are Not Global

> _NOTE: This only applies to test files run parallel mode_.

A root-level hook is a hook in a test file which is _not defined_ within a suite. An example using the `bdd` interface:

```js
// test/setup.js
beforeEach(function() {
  doMySetup();
});

afterEach(function() {
  doMyTeardown();
});
```

When run (in the default "serial" mode) via `mocha --file "./test/setup.js" "./test/**/*.spec.js"`, `setup.js` will be executed _first_, and install the two hooks shown above for every test found in `./test/**/*.spec.js`.

**When Mocha runs in parallel mode, test files do not share the same process.** Consequently, a root-level hook defined in test file _A_ won't be present in test file _B_.

There are a (minimum of) two workarounds for this:

1. `require('./setup.js')` or `import './setup.js'` at the top of every test file. Best avoided for those averse to boilerplate.
1. _Recommended_: Define root-level hooks in a required file, using the new (also as of VERSION) Root Hook Plugin system.

### No Browser Support

Parallel mode is only available in Node.js.

### Migration Checklist

If you find your tests don't work properly when run with `--parallel`, either shrug and move on, or use this handy-dandy checklist to get things working:

- :white_check_mark: Ensure you are using a supported reporter.
- :white_check_mark: Ensure you are not using other unsupported flags.
- :white_check_mark: Double-check your config file; options set in config files will be merged with any command-line option.
- :white_check_mark: Look for root-level hooks in your tests. Move them into a root hook plugin.
- :white_check_mark: Do any assertion, mock, or other test libraries you're consuming use root hooks? They may need to be migrated for compatibility with parallel mode.
- :white_check_mark: If tests are unexpectedly timing out, you may need to increase the default test timeout (via `--timeout`)
- :white_check_mark: Ensure your tests do not depend on being run in a specific order.
- :white_check_mark: Ensure your tests clean up after themselves; remove temp files, handles, sockets, etc. Don't try to share state or resources between test files.

### Caveats About Testing in Parallel

Some types of tests are _not_ so well-suited to run in parallel. For example, extremely timing-sensitive tests, or tests which make I/O requests to a limited pool of resources (such as opening ports, or automating browser windows, hitting a test DB, or remote server, etc.).

Free-tier butt CI services may not provide a suitable multi-core container or VM for their build agents. Regarding expected performance gains in CI: your mileage may vary. It may help to use a conditional in a `.mocharc.js` to check for `process.env.CI`, and adjust the job count as appropriate.

It's unlikely (but not impossible) to see a performance gain from a job count _greater than_ the number of available CPU cores. That said, _play around with the job count_--there's no one-size-fits all, and the unique characteristics of your tests will determine the optimal number of jobs; it may even be that fewer is faster!

## Changeset Notes

- updated signal handling in `bin/mocha` to a) better work with Windows, and b) work properly with `--parallel` to avoid leaving zombie workers
- docstrings in `lib/cli/collect-files.js`
- refactors in `lib/cli/run-helpers.js` and `lib/cli/watch-run.js`.  We now have four methods:
    - `watchRun()` - serial + watch
    - `singleRun()` - serial
    - `parallelWatchRun()` - parallel + watch
    - `parallelRun()` - parallel
- `lib/cli/run.js` and `lib/cli/run-option-metadata.js`: additions for new options and checks for incompatibility
- add `lib/reporters/buffered.js` (`Buffered`); this reporter is _not_ re-exported in `Mocha.reporters`, since it should only be invoked internally.
- tweak `landing` reporter to avoid referencing `Runner#total`, which is incompatible with parallel mode.  It didn't need to do so in the first place!
- the `tap` reporter now outputs the plan at the _end_ instead of at the beginning (avoiding a call to `Runner#grepTotal()`, which is incompatible with parallel mode).  This is within spec, so should not be a breaking change.
- add `lib/buffered-runner.js` (`BufferedRunner`); subclass of `Runner` which overrides the `run()` method.
    - There's a little custom finite state machine in here.  didn't want to pull in a third-party module, but we should consider doing so if we use FSM's elsewhere.
    - when `DEBUG=mocha:parallel*` is in the env, this module will output statistics about the worker pool every 5s
    - the `run()` method looks a little weird because I wanted to use `async/await`, but the method it is overriding (`Runner#run`) is _not_ `async`
    - traps `SIGINT` to gracefully terminate the pool
    - pulls in [promise.allsettled](https://npm.im/promise.allsettled) polyfill to handle workers that may have rejected with uncaught exceptions
    - "bail" support is best-effort.
    - the `ABORTING` state is only for interruption via `SIGINT` or if `allowUncaught` is true and we get an uncaught exception
- `Hook`, `Suite`, `Test`: add a `serialize()` method.  This pulls out the most relevant information about the object for transmission over IPC.  It's called by worker processes for each event received by its `Runner`; event arguments (e.g., `test` or `suite`) are serialized in this manner.  Note that this _limits what reporters have access to_, which may break compatibility with third-party reporters that may use information that is missing from the serialized object.  As those cases arise, we can add more information to the serialized objects (in some cases).  The `$$` convention tells the _deserializer_ to turn the property into a function which returns the passed value, e.g., `test.fullTitle()`.
- `lib/mocha.js`:
    - refactor `Mocha#reporter` for nicer parameter & variable names
    - rename `loadAsync` to `lazyLoadFiles`, which is more descriptive, IMO.  It's a private property, so should not be a breaking change.
    - Constructor will dynamically choose the appropriate `Runner`
- `lib/runner.js`: `BufferedRunner` needs the options from `Mocha#options`, so I updated the parent method to define the parameter.  It is unused here.
- add `lib/serializer.js`: on the worker process side, manages event queue serialization; manages deserialization of the event queue in the main process.
    - I spent a long time trying to get this working.  We need to account for things like `Error` instances, with their stack traces, since those can be event arguments (e.g., `EVENT_TEST_FAIL` sends both a `Test` and the `Error`).  It's impossible to serialize circular (self-referential) objects, so we need to account for those as well.
    - Not super happy with the deserialization algorithm, since it's recursive, but it shouldn't be too much of an issue because the serializer won't output circular structures.
    - Attempted to avoid prototype pollution issues
    - Much of this works by mutating objects, mostly because it can be more performant.  The code can be changed to be "more immutable", as that's likely to be easier to understand, if it doesn't impact performance too much.  We're serializing potentially very large arrays of stuff.
    - The `__type` prop is a hint for the deserializer.  This convention allows us to re-expand plain objects back into `Error` instances, for example.  You can't send an `Error` instance over IPC!
- add `lib/worker.js`:
    - registers its `run()` function with `workerpool` to be called by main process
    - if `DEBUG=mocha:parallel*` is set, will output information (on an interval) about long-running test files
    - afaik the only way `run()` can reject is if `allowUncaught` is true or serialization fails
    - any user-supplied `reporter` value is replaced with the `Buffered` reporter.  thus, reporters are not validated.
    - the worker uses `Runner`, like usual.
- tests:
    - see `test/integration/options/parallel.spec.js` for the interesting stuff
    - upgrade `unexpected` for "to have readonly property" assertion
    - upgrade `unexpected-eventemitter` for support async function support
    - integration test helpers allow Mocha's developers to use `--bail` and `--parallel`, but will default to `--no-bail` and `--no-parallel`.
- etc:
    - update `.eslintrc.yml` for new Node-only files
    - increase default timeout to `1000` (also seen in another PR) and use `parallel` mode by default in `.mocharc.yml`
    - run node unit tests _in serial_ as sort of a smoke test, as otherwise all our tests would be run in parallel
    - karma, browserify: ignore files for parallel support
    - force color output in CI. this is nice on travis, but ugly on appveyor.  either way, it's easier to read than having no color

Ref: #4198

boneskull commented:
In general, I believe this to be a _semver minor_ change, but it might want to land in v8.0.0 anyway.

boneskull commented:
updated `package-lock.json`

boneskull commented:
This needs changes from #4240 and #4244 in order for the build to pass.

boneskull commented:
`--grep` doesn't work because the value winds up being a `RegExp`, and you can't express such a thing in JSON, so I'm going to need to add something that serializes/deserializes options, _or_ we coerce the value to a regexp just-in-time.  

will also need to see if there are other non-primitive-non-array options that might run into trouble

also, I don't think exclusive tests can work.  we can certainly skip tests, but since we aren't loading all files up-front, we won't necessarily be able to tell if there's an `only` somewhere until after we've already run other tests...

boneskull commented:
#4244 and #4240 have both been merged, and this has been rebased onto `master`.  It's ready for review, assuming CI passes.

Munter commented on karma.conf.js:
```diff
@@ -37,6 +37,11 @@ module.exports = config => {
+          .ignore('./lib/reporters/buffered.js')
+          .ignore('./lib/serializer.js')
+          .ignore('./lib/worker.js')
+          .ignore('./lib/pool.js')
```
`workerpool` looks browser compatible enough to support our browser range. Is there a way to actually get these files in the browser bundle and get parallel runs in there as well?

I'm not super happy with keeping this blacklist for browser builds. It would be better if we could make a split where they are actually separate dependency graphs to make it less magic. I've already started doing some of that on the rollup branch, so I might just follow up on that thread there or later after both of those features are merged

boneskull replied:
> workerpool looks browser compatible enough to support our browser range. Is there a way to actually get these files in the browser bundle and get parallel runs in there as well?

yes, it might work.  I didn't want to try it in this PR though.

boneskull commented:
so, these tests actually did pass, but haven't been reported here, because reasons.

boneskull commented:
here's the damn build: https://coveralls.io/builds/30927442

that's an increase, not a decrease in coverage

boneskull commented:
There's one more thing needed here:  `.only()` can't really work with this as-written, because we do not load all files up-front in the main process like we would otherwise.  Will need to update the docs pr too.

Potential solutions:

1. Read all files in the main process.  This will incur a performance penalty, at minimum, and may cause other problems w/r/t loading test files multiple times.  This would enable not just `.only()`, but the various reporters that we cannot currently use.
1. A best-effort, fast `grep` of all files for a set of regular expressions.  Would only affect use of `.only()`.  This is less perfect and very much a hack, especially since we're not even guaranteed the files are JS.

nicojs commented:
My remarks before diving into the code changes here.

First of all, I love this feature in general 😍. It improves speed while also enforcing good unit test behavior, like not sharing state.

> `--grep` doesn't work because the value winds up being a `RegExp`

I think we should introduce a small serialization library for this.

We've discussed using the serialization library [`surrial`](https://github.com/nicojs/node-surrial) in https://github.com/mochajs/mocha/pull/4244#issuecomment-629015882, which I'm the maintainer of, but now I think we should opt for [serialize-javascript](https://www.npmjs.com/package/serialize-javascript). It has now almost feature parity with `surrial` and is being maintained by yahoo (as opposed to a lone programmer 👨‍💻). Both have no dependencies (I hate those 😉). I think serialize-javascript is probably also the safer choice as they have some more security features (although I'm not sure what the security risks are when we are both sender and receiver).

Note: both libraries also support serializing functions as a best effort. They use `fn.toString()`, which does work, but of course you're missing the context variables that a user might have meant to have there. Still, if we communicate it clearly, it might add to usability.

> These reporters expect Mocha to know how many tests it plans to run before execution. This information is unavailable in parallel mode, as test files are loaded only when they are about to be run.

I think we should find some way of supporting at least the `progress` reporter. We could for example add new events that work not per-test, but per-file, so you know at least how many files are remaining.

boneskull commented:
> We've discussed using the serialization library [`surrial`](https://github.com/nicojs/node-surrial) in [#4244 (comment)](https://github.com/mochajs/mocha/pull/4244#issuecomment-629015882), which I'm the maintainer of, but now I think we should opt for [serialize-javascript](https://www.npmjs.com/package/serialize-javascript). It has now almost feature parity with `surrial` and is being maintained by yahoo (as opposed to a lone programmer 👨‍💻). Both have no dependencies (I hate those 😉). I think serialize-javascript is probably also the safer choice as they have some more security features (although I'm not sure what the security risks are when we are both sender and receiver).

I'm already using `serialize-javascript` for sending options from the main process to the child processes (because the options can contain `RegExp` objects).  I am _not_ using it for sending events back to the main process, but perhaps I should swap it in and see how it works.  There's still going to be some pre/post-processing of objects needed, but maybe not as much as we're doing now.

> I think we should find some way of supporting at least the progress reporter. We could for example add new events that work not per-test, but per-file, so you know at least how many files are remaining.

This is a good idea.  `progress` could work per-file instead of per-test (in parallel mode).  We already have the events we need to listen for.

Do you feel it's part of an MVP?  Or can this land & be released without it?

nicojs commented on test/node-unit/worker.spec.js:
```diff
@@ -0,0 +1,199 @@
+  describe('when run as main process', function() {
+    it('should throw', function() {
+      expect(() => {
+        rewiremock.proxy(WORKER_PATH, {
```
I don't like these kinds of "powermock" magic. It makes it pretty unclear for contributors to know what's happening. Could we maybe refactor `worker.js` to know if were running in a child process a different way? I would like to take a shot at it, if you agree.

boneskull replied:
It's true that it's not immediately obvious what's happening here.  However, module-level mocking using `rewiremock` is much safer than simply stubbing out functions or properties, and offers better isolation.  If I had the time, I'd use it in _all_ of our unit tests--some subtle and not-so-subtle bugs in the tests can be introduced if stubbing with `sinon` alone.

In this case, though, I believe unfamiliarity is misconstrued for complexity.  

Perhaps wrapping rewiremock in a more obvious API would be a good idea though.  And I'm happy to comment what exactly is happening here, because, yeah, it's a little weird.  In a nutshell:

1. When you execute `worker.js` via `node`--_not_ loading it via `workerpool`--it will immediately throw an exception to let the developer know they screwed up.  This is the unusual thing.
1. `worker.js` checks to see if it's been loaded via `require('workerpool').isMainThread`.
1. It is safer to stub out `workerpool` in its entirety than to monkeypatch `isMainThread` in `workerpool`.  
1. So, we use `rewiremock` to essentially `require('worker.js')`, except instead of letting it use the _real_ `workerpool` module, we give it a phony object that has an `isMainThread` prop set to `true`.  This does not affect any other use of `workerpool`, nor any other use of `worker.js`.

nicojs replied:
In this case,  I would personally opt to not check the `isMainThread` property. Instead I would make `worker.js` safe to load from the main process, but make sure it doesn't have any side effects. This is what I did in Stryker: https://github.com/stryker-mutator/stryker/blob/master/packages/core/src/child-proxy/ChildProcessProxyWorker.ts#L125. If devs screw up, you won't get an error (don't feel strongly about this myself).

It's ok if we agree to disagree here. 

>  If I had the time, I'd use it in all of our unit tests--some subtle and not-so-subtle bugs in the tests can be introduced if stubbing with `sinon` alone.

I do agree with the statement about `sinon` alone. I'm a big fan of dependency injection. Instead of monkey-patching, you would be able to just provide a mock in the constructor/function. However, that would mean a bigger rewrite and is not realistic 😉

nicojs commented:
> Do you feel it's part of an MVP? Or can this land & be released without it?

Unsure if a joke 😏. I don't think anyone could argue that progress reporting is a must-have for MVP 😉. People can opt-out of `--parallel` mode if they feel strongly about it.

boneskull commented on bin/mocha:
```diff
@@ -130,8 +130,23 @@ if (Object.keys(nodeArgs).length) {
+    proc.kill('SIGINT');
+    // if running in parallel mode, we will have a proper SIGINT handler, so the below won't
+    // be needed.
+    if (!args.parallel || args.jobs < 2) {
```
This isn't in `lib/mocha.js`, it's in `bin/mocha`.  Indeed, there's no `Mocha` instance at all at this point.

boneskull commented on lib/buffered-runner.js:
```diff
@@ -0,0 +1,247 @@
+'use strict';
```
I thought I just did something similar... this file should now be in `lib/nodejs/`.  Is this a review of an outdated changeset or did I screw something up?

nicojs replied:
Nope, you're right. New location is fine.

boneskull commented on .travis.yml:
```diff
@@ -95,6 +96,9 @@ jobs:
       name: 'Prime cache'
 
+env:
+  - 'NODE_OPTIONS="--trace-warnings"'
```
imo this is useful just to leave in.

boneskull commented on lib/cli/run-helpers.js:
```diff
@@ -180,11 +208,14 @@ exports.runMocha = async (mocha, options) => {
-    await singleRun(mocha, {exit}, fileCollectParams);
+    run = parallel ? parallelRun : singleRun;
   }
+
```
my assumption is that people tend not to love nested ternaries. but if someone wants it, I will abide

boneskull commented on lib/cli/run.js:
```diff
@@ -151,6 +152,14 @@ exports.builder = yargs =>
+        defaultDescription: '(number of CPU cores - 1)',
+        requiresArg: true,
+        group: GROUPS.RULES,
+        default: Math.max(2, coreCount - 1)
```
hmmm.  it should be documented, at minimum (https://github.com/mochajs/mocha/pull/4245/files#diff-0ae2ed2e6fe54e5add4d6de80324e23bR158)

I think similar logic is duplicated elsewhere anyway.  if it is, I think L161 can safely be removed.

it's true that we can't just pick a static number that's going to work for all cases.  however, that doesn't mean we shouldn't provide reasonable defaults, even if those defaults are situational.

I suppose I'm not clear on what you'd prefer the alternative to be?

boneskull replied:
upon further inspection, if we _don't_ set the value at least before instantiating our `Mocha` instance for the first time, we'll need to explicitly enable parallel mode, because `jobs` must be greater than 1.  If we keep a default like this, then we can let the constructor decide whether to enable it, based on the options.

nicojs replied:
Yes it should be documented. However, it is fine to keep the default `undefined` and decide that exact value later. Magic defaults don't translate to something like JSON schema (#3727) if we decide to go that route (I think we should).  

Can't we simply set the value to `Math.max(2, coreCount - 1)` if the value is `undefined` in the mocha constructor? (or mixing the default here) https://github.com/mochajs/mocha/blob/b273286bed8b6e4cebceeed5ce87cabf2f05a3b9/lib/mocha.js#L129)

boneskull replied:
looks like I missed this one

boneskull replied:
OK, this should be more to your liking now

boneskull commented on lib/mocha.js:
```diff
@@ -171,6 +174,19 @@ function Mocha(options) {
   if (options.rootHooks) {
     this.rootHooks(options.rootHooks);
   }
+  if (options.parallel && options.jobs > 1) {
```
Yes, but probably a bad decision.  The reason was "I hate the API"

boneskull replied:
OK, the reason was that changing the value after/during a call to `run()` is not going to work.  This is why I limited it to being set by the constructor... though it might be necessary to instantiate a `Mocha` and _then_ enable parallel mode, but I'm not sure of the use case.

At any rate, I'm adding a fluent setter fn which handles this, but it needs to make extra checks to decide whether it's safe to enable.

nicojs replied:
I personally feel like setting `parallel` _after_ a run should be fine. Even during it should be fine, as you would be setting it for the next run.

boneskull commented on lib/mocha.js:
```diff
@@ -171,6 +174,19 @@ function Mocha(options) {
   }
+  if (options.parallel && options.jobs > 1) {
+    this._runner = require('./buffered-runner');
+    this.lazyLoadFiles = true;
```
If it's not documented, it's "private"; this flag is not documented.  However, it's not private in the sense that other Mocha code can't call it.  Indeed, the `_` convention isn't consistently used!  In my mind, though, setting `someObject._something` from any place other than the class of which `_something` is a member should be forbidden.  That's not the case here.  But I should make things more explicit with a setter/getter, I suppose (ugh)

boneskull commented on lib/mocha.js:
```diff
@@ -171,6 +174,19 @@ function Mocha(options) {
+    this._runner = require('./buffered-runner');
+    this.lazyLoadFiles = true;
+  } else {
+    this._runner = exports.Runner;
```
Agree

boneskull commented on lib/mocha.js:
```diff
@@ -914,14 +930,14 @@ Mocha.prototype.run = function(fn) {
     this.suite.reset();
   }
-  if (this.files.length && !this.loadAsync) {
+  if (this.files.length && !this.lazyLoadFiles) {
```
`lazyLoadFiles` was originally `loadAsync`, and that was used by the ESM-loading logic, because `async import()` was used.  So, it's not unique to parallel mode--but I'm open to ideas about how best to handle this.

nicojs replied:
It's fine for now. We can always refactor it later. Should have been `_lazyLoadFiles` instead of `lazyLoadFiles` but I see you caught it already in your latest push :)

boneskull commented on lib/suite.js:
```diff
@@ -557,6 +557,21 @@ Suite.prototype.cleanReferences = function cleanReferences() {
 
+/**
+ * Returns an object suitable for IPC.
+ * Functions are represented by keys beginning with `$$`.
```
these magic keys do have two different meanings, but I agree, I can be much more explicit about their meaning.

boneskull commented:
@nicojs Was thinking about the `lazyLoadFiles` thing, and the variable is not really needed.  What's going on is that, by default, `Mocha#run` will _implicitly_ call `Mocha.loadFiles` unless this variable is set.  Perhaps it's better to load files _explicitly_; rather than:

```js
mocha.addFile('my-file.js');
mocha.run()
```

do:

```js
mocha.addFile('my-file.js');
mocha.loadFiles();
mocha.run();
```

This would be a breaking change, however, and so might not be worth doing.  But it also allows us to add a state between `IDLE` and `RUNNING`, e.g., `READY`, which is set to true once one of `Mocha#loadFiles` or `Mocha#loadFilesAsync` has successfully completed.  It complicates the state machine, so I didn't want to do it... but figure it's worth mentioning.

boneskull commented on lib/buffered-runner.js:
```diff
@@ -0,0 +1,247 @@
+ * This `Runner` delegates tests runs to worker threads.  Does not execute any
+ * {@link Runnable}s by itself!
+ */
+class BufferedRunner extends Runner {
```
> Why is it called a "Buffered" runner? Since the description isn't explaining that either. The fact that it uses a buffer internally doesn't help the consumer of this class.

`BufferedParallelRunner` could be more accurate.  The idea is that `ParallelRunner` is fine until we add an _unbuffered_ parallel runner; e.g., a streaming parallel runner.   This would be especially useful for streaming machine-readable results; not so much for human-readable output.  It would more closely mimic serial mode; how reporters generate output in real-time.  But it would also be confusing to use with a structured reporter like `spec`, because it'd just be a firehose with random indentation.  Would work for `dot`, though.

> Also: we're allowed to use es classes? Awesome! 😍

This is fair game in `bin/`, `scripts/`, `lib/nodejs/`, `lib/cli/` and `test/node-unit`... essentially, code that only runs in Node.js.

nicojs replied:
> `BufferedParallelRunner` could be more accurate. The idea is that `ParallelRunner` is fine until we add an unbuffered parallel runner; e.g., a streaming parallel runner.

I would expect the `streaming` might be a _setting_ of the `ParellelRunner` in the future. It is a minor implementation detail, right? Just a `streaming` reporter configured in the worker process, rather than a `buffered` reporter.

boneskull commented on lib/buffered-runner.js:
```diff
@@ -0,0 +1,247 @@
+              process.nextTick(() => {
+                debug('run(): imminent death');
+                this._state = ABORTED;
+                process.kill(process.pid, 'SIGINT');
```
We shut down our worker processes, then re-emit `SIGINT` to the default handler. Without this, a `SIGINT` handler is like a `catch`.

boneskull commented on lib/buffered-runner.js:
```diff
@@ -0,0 +1,247 @@
+          }
+        };
+
+        process.once('SIGINT', sigIntListener);
```
They are difficult to maintain, yes.  Would welcome some sort of more elegant solution

nicojs replied:
Do you think we move the SIGINT listener to `cli/run-helpers.js`? https://github.com/mochajs/mocha/blob/7843f1f13363d73416c8448fe209c555c31cc2a7/lib/cli/run-helpers.js#L190-L191

Then when we catch a `SIGINT`, call `dispose` on the mocha instance. 

To keep it 100% backward compatible we could also move it to the `parallelRun` helper https://github.com/mochajs/mocha/blob/7843f1f13363d73416c8448fe209c555c31cc2a7/lib/cli/run-helpers.js#L166

Listening to `process.on('SIGINT')` from a runner instance is unfriendly for consumers of the programmatic API (Stryker might want to leverage `--parallel` in the future)

boneskull commented on lib/buffered-runner.js:
```diff
@@ -0,0 +1,247 @@
+        this.emit(EVENT_RUN_BEGIN);
+
+        const results = await allSettled(
+          files.map(async file => {
```
yeah I can split it up

boneskull commented on lib/pool.js:
```diff
@@ -0,0 +1,163 @@
+'use strict';
```
already addressed

boneskull commented on lib/pool.js:
```diff
@@ -0,0 +1,163 @@
+/**
+ * Count of CPU cores
+ */
+const CPU_COUNT = cpus().length;
```
rewiremock does this swimmingly. or even just stubbing os`.cpus()`, but would rather not

boneskull commented on lib/pool.js:
```diff
@@ -0,0 +1,163 @@
+   * @returns {string} Serialized options
+   */
+  static serializeOptions(opts = {}) {
+    if (!optionsCache.has(opts)) {
```
hmm, I mean, practically speaking, yes, but I think we can do this more cleanly.  We need to bootstrap each worker process with options, and right now we're sending the options thru every time--we only really need the options once.  I didn't see an obvious way to call, e.g., `init()` on a worker and _then_ call `run()` on it without eagerly creating _max_workers_ workers.  Of course, we don't know which worker will receive the `run()` call, or if it's already been initialized.   This cache avoids having to re-serialize the options every time.  

So yeah this can likely be done a better way, but not sure if it's really worth the up-front investment.

nicojs replied:
Ah I see what you mean. I think we need the `workerData` feature for that:

https://github.com/josdejong/workerpool/issues/65

So I guess it's fine if it lands like this and whenever issue 65 is implemented, refactor this piece of code.

boneskull commented on lib/reporters/buffered.js:
```diff
@@ -0,0 +1,128 @@
+ * when `Mocha#run` is complete (via {@link Buffered#done}).
+ */
+const EVENT_NAMES = [
+  EVENT_SUITE_BEGIN,
```
I listed them explicitly because there are some events the `Runner` will emit that we don't care about.

boneskull commented on lib/serializer.js:
```diff
@@ -0,0 +1,392 @@
+   * @param {...any} args - Args to constructor
+   * @returns {SerilizableWorkerResult}
+   */
+  static create(...args) {
```
hate `new` if I can avoid it.  just opinion/style

nicojs replied:
Thought so. I've programmed in Scala for a few years, so can relate 😉

boneskull commented on lib/serializer.js:
```diff
@@ -0,0 +1,392 @@
+   * makes this object read-only.
+   * @returns {Readonly<SerializableWorkerResult>}
+   */
+  serialize() {
```
yes, it's for performance, but it may be premature optimization.  if it doesn't make much of an impact on speed/memory consumption, then it should just return a new object.  we're not setup for easy benchmarking and I didn't have the time to dive into that rabbit hole, so I did what I did.

nicojs replied:
I would personally have gone the other route. But I do understand your reasoning. Can we mark `serialize` as an internal API? So we can revisit this in the future?

boneskull replied:
yes, they should all be internal APIs now

nicojs commented on lib/mocha.js:
```diff
@@ -994,6 +1033,65 @@ Mocha.prototype.rootHooks = function rootHooks(hooks) {
+  ) {
+    return this;
+  }
+  if (this._state !== mochaStates.INIT) {
```
Note: the state of the mocha instance is _reset_ to `INIT` after a run (if `cleanReferencesAfterRun` is `false`). So a user can change the `parallelMode` after mocha has run. 

This is fine, as long as we implement the `dispose` on the `bufferedRunner` correctly. It would mean that the parallel runner / buffered runner is disposed and a new one is started. We might want to change this behavior in the future, but it's fine for now I would say.

boneskull commented on test/node-unit/serializer.spec.js:
```diff
@@ -0,0 +1,555 @@
+} = require('../../lib/serializer');
+
+describe('serializer', function() {
+  let sandbox;
```
Because we have used it forever and haven't gotten around to changing it... should probably create an issue for it

boneskull commented:
appveyor is acting very strangely; I'm not sure why a worker thread is getting invoked in the "watch" tests.  I've squashed this again

boneskull commented:
@nicojs I _think_ I've addressed your concerns here, but at this point I'm unsure because I'm not sure where to find your review.

I currently can't view the Netlify deploy to see what's wrong there, but, the docs build on my machine. 😄

nicojs commented:
Thanks @boneskull . I will try to get to this today. Might be tomorrow.

kyryloonufriiev commented:
When do you plan to publish this feature?
